File: csharp/Behavioral/21-TemplateMethod/Report/JobInspectionReport.cs
namespace TemplateMethod;

public abstract class JobInspectionReport
{
    protected string JobId { get; private set; } = string.Empty;
    protected string Inspector { get; private set; } = string.Empty;
    protected readonly List<string> Findings = new();

    // ══════════════════════════════════════════════════════
    // THE TEMPLATE METHOD
    // Not virtual — subclasses cannot change the sequence,
    // only the individual steps.
    // ══════════════════════════════════════════════════════
    public void GenerateReport(string jobId, string inspector)
    {
        JobId     = jobId;
        Inspector = inspector;
        Findings.Clear();

        OpenReport();              // concrete — same for all
        VerifySafety();            // abstract — subclass must implement
        PerformInspection();       // abstract — subclass must implement
        RecordFindings();          // abstract — subclass must implement
        CaptureAdditionalPhotos(); // hook     — optional override
        AttachComplianceNote();    // abstract — subclass must implement
        SignOff();                 // concrete — same for all
    }

    // ── Concrete steps — identical for every trade ─────────

    private void OpenReport()
    {
        Console.WriteLine("\n╔══════════════════════════════════════════════════");
        Console.WriteLine($"║ INSPECTION REPORT — {TradeName}");
        Console.WriteLine("╠══════════════════════════════════════════════════");
        Console.WriteLine($"║ Job ID    : #{JobId}");
        Console.WriteLine($"║ Inspector : {Inspector}");
        Console.WriteLine($"║ Opened    : {DateTime.Now:dd MMM yyyy HH:mm}");
        Console.WriteLine("╠══════════════════════════════════════════════════");
    }

    private void SignOff()
    {
        Console.WriteLine("╠══════════════════════════════════════════════════");
        Console.WriteLine($"║ Findings recorded : {Findings.Count}");
        Console.WriteLine($"║ Signed off by     : {Inspector}");
        Console.WriteLine($"║ Closed            : {DateTime.Now:dd MMM yyyy HH:mm}");
        Console.WriteLine("╚══════════════════════════════════════════════════");
    }

    // ── Abstract steps — subclasses MUST implement ─────────

    protected abstract string TradeName { get; }

    protected abstract void VerifySafety();
    protected abstract void PerformInspection();
    protected abstract void RecordFindings();
    protected abstract void AttachComplianceNote();

    // ── Hook — optional, empty by default ──────────────────

    protected virtual void CaptureAdditionalPhotos()
    {
        // Default: no additional photos required
    }

    // ── Helper available to all subclasses ─────────────────

    protected void AddFinding(string finding)
    {
        Findings.Add(finding);
        Console.WriteLine($"║   • {finding}");
    }
}

File: typescript/Behavioral/21-TemplateMethod/Report/JobInspectionReport.ts
export abstract class JobInspectionReport {
  protected jobId: string = "";
  protected inspector: string = "";
  protected findings: string[] = [];

  // ══════════════════════════════════════════════════════
  // THE TEMPLATE METHOD
  // TypeScript has no 'sealed' — convention is to document
  // that subclasses must not override this method.
  // ══════════════════════════════════════════════════════
  generateReport(jobId: string, inspector: string): void {
    this.jobId     = jobId;
    this.inspector = inspector;
    this.findings  = [];

    this.openReport();              // concrete
    this.verifySafety();            // abstract
    this.performInspection();       // abstract
    this.recordFindings();          // abstract
    this.captureAdditionalPhotos(); // hook
    this.attachComplianceNote();    // abstract
    this.signOff();                 // concrete
  }

  // ── Concrete steps — identical for every trade ─────────

  private openReport(): void {
    console.log("\n╔══════════════════════════════════════════════════");
    console.log(`║ INSPECTION REPORT — ${this.tradeName}`);
    console.log("╠══════════════════════════════════════════════════");
    console.log(`║ Job ID    : #${this.jobId}`);
    console.log(`║ Inspector : ${this.inspector}`);
    console.log(`║ Opened    : ${this.timestamp()}`);
    console.log("╠══════════════════════════════════════════════════");
  }

  private signOff(): void {
    console.log("╠══════════════════════════════════════════════════");
    console.log(`║ Findings recorded : ${this.findings.length}`);
    console.log(`║ Signed off by     : ${this.inspector}`);
    console.log(`║ Closed            : ${this.timestamp()}`);
    console.log("╚══════════════════════════════════════════════════");
  }

  private timestamp(): string {
    return new Date().toLocaleString("en-GB", {
      day: "2-digit",
      month: "short",
      year: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  }

  // ── Abstract steps — subclasses MUST implement ─────────

  protected abstract get tradeName(): string;

  protected abstract verifySafety(): void;
  protected abstract performInspection(): void;
  protected abstract recordFindings(): void;
  protected abstract attachComplianceNote(): void;

  // ── Hook — optional, empty by default ──────────────────

  protected captureAdditionalPhotos(): void {
    // Default: no additional photos required
  }

  // ── Helper available to all subclasses ─────────────────

  protected addFinding(finding: string): void {
    this.findings.push(finding);
    console.log(`║   • ${finding}`);
  }
}
